src/input/json_to_storage.cc

```diff
@@ -93,106 +231,362 @@ bool JSONPath::get_ref_from_head(string & ref_address)
+        return j_array.size();
     }
 
-    for(vector< pair<int, string> >::const_iterator it = path_.begin()+1; it != path_.end(); ++it) {
-        if ( it->first < 0 ) {
-            stream << "/" << it->second;
-        } else {
-            stream << "/" << it->first;
+    return -1;
+}
+
+
+
+bool PathJSON::is_map_type() const {
+	return head()->type() == json_spirit::obj_type;
+}
+
+
+
+bool PathJSON::is_sequence_type() const {
+	return head()->type() == json_spirit::array_type;
+}
+
+
+
+PathJSON * PathJSON::clone() const {
+	return new PathJSON(*this);
+}
+
+
+
+bool PathJSON::has_descendent_index(bool value_at_declaration) {
+	if ( this->is_map_type() ) {
+
+		PathBase *type_path = this->clone();
+		if ( !type_path->down("TYPE") ) {
+            if ( !value_at_declaration ) {
+                THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("Missing key 'TYPE' in AbstractRecord.")
+                	<< JSONToStorage::EI_ErrorAddress(this) << JSONToStorage::EI_Format("JSON") );
+            } else { // auto conversion
+            	return false;
+            }
+		} else {
+			return true;
+		}
+	} else {
+        if ( !value_at_declaration ) {
+            THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'JSON object', but we found: ")
+                << JSONToStorage::EI_ErrorAddress(this) << JSONToStorage::EI_Format("JSON") << JSONToStorage::EI_JSON_Type( this->get_node_type() ) );
+        } else { // auto conversion
+        	return false;
         }
+	}
+
+	return false;
+}
+
+
+
+std::ostream& operator<<(std::ostream& stream, const PathJSON& path) {
+    path.output(stream);
+    return stream;
+}
+
+
+/********************************************
+ * Implementation of PathYAML
+ */
+
+PathYAML::PathYAML(istream &in)
+: PathBase()
+{
+	PathYAML::Node root_node = YAML::Load( in );
+    nodes_.push_back( new Node(root_node) );
+}
+
+
+bool PathYAML::down(unsigned int index) {
+    const Node & head_node = *( nodes_.back() );
+    ASSERT(head_node.IsSequence(), "Head node must be of type Array.\n");
+
+    if ( index >= head_node.size() ) return false;
+    path_.push_back( make_pair( index, string("") ) );
+    nodes_.push_back( new Node( head_node[index] ) );
+
+    return (nodes_.back() != NULL);
+}
+
+
+bool PathYAML::down(const string& key) {
+    const Node & head_node = *( nodes_.back() );
+    ASSERT(head_node.IsMap(), "Head node must be of type Record.\n");
+
+    if (head_node[key]) {
+    	path_.push_back( make_pair( (int)(-1), key) );
+    	nodes_.push_back( new Node( head_node[key] ) );
+    } else {
+        return false;
     }
+    return (nodes_.back() != NULL);
 }
 
 
+void PathYAML::up() {
+    if (path_.size() > 1) {
+        path_.pop_back();
+        nodes_.pop_back();
+    }
+}
+
 
-string JSONPath::str() {
-    stringstream ss;
-    output(ss);
-    return ss.str();
+void PathYAML::go_to_root() {
```

Comment:
Again memory leak. I propose to implement go_to_root in PathBase using up() and level().